### PR TITLE
Implement C++ ModuleDict

### LINKED ISSUE #45896: C++ ModuleDict
Is ModuleDict still slated to be added to the C++ api?

It looks like pr #28652, opened in Oct 2019 was close to completion,
then was never finished.

We can take the code in this pr's _moduledict.h_ for now,
but wanted to check if the container is still destined for the C++ api at some point.

Thanks.

### PR DESCRIPTION
**Pull request by ejguan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47707 Implement C++ ModuleDict**

Fixes #45896
- [x] API
* constructor(OrderedDict/vector of pair)
* clear()
* items()
* keys()
* pop(string key)
* update(OrderedDict/vector of pair/ModuleDict)
* values()
* Iterator: begin/end
* size()
* empty()
* contains(string key)
* clone()
* pretty_print(stream)
* operator [string key]
* at(string key)
- [x] Add tests

Differential Revision: [D24872641](https://our.internmc.facebook.com/intern/diff/D24872641)

### CODE REVIEW

File: torch/csrc/api/include/torch/nn/modules/container/moduledict.h [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+namespace nn {
+
+class ModuleDictImpl : public Cloneable<ModuleDictImpl> {
+ public:
+  using Iterator = torch::OrderedDict<std::string, std::shared_ptr<Module>>::Iterator;
+  using ConstIterator = torch::OrderedDict<std::string, std::shared_ptr<Module>>::ConstIterator;
+
+  ModuleDictImpl() = default;
+
+  /// Constructs the `ModuleDict` from a list of string-Module paris.
```
Comment (zou3519) [inline torch/csrc/api/include/torch/nn/modules/container/moduledict.h:18]: nit: pairs

File: torch/csrc/api/include/torch/nn/modules/container/moduledict.h [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    auto clone = std::make_shared<ModuleDictImpl>();
+    for (const auto& module : modules_) {
+      clone->insert(module.key(), module.value()->clone(device));
+    }
+    return clone;
+  }
+
+  /// `reset()` is empty for `ModuleDict`, since it does not have parameters of
+  /// its own.
+  void reset() override {}
```
Comment (zou3519) [inline torch/csrc/api/include/torch/nn/modules/container/moduledict.h:97]: @glaringlee how is reset() used in the C++ API? If it doesn't make sense to call reset() for ModuleDict, would it be better to raise an error?

File: torch/csrc/api/include/torch/nn/modules/container/moduledict.h [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+  /// Removes and returns the value associated with the given `key`.
+  /// Throws an exception if no such key is stored in the `ModuleDict`.
+  /// Check contains(key) before for a non-throwing way of access.
+  std::shared_ptr<Module> pop(const std::string& key) {
+    auto module = modules_[key];
+    modules_.erase(key);
+    return module;
+  }
+
+  /// Updated the `ModuleDict` with key-value paris from vector of key-module pairs
```
Comment (zou3519) [inline torch/csrc/api/include/torch/nn/modules/container/moduledict.h:149]: nit: paris -> pairs

File: test/cpp/api/moduledict.cpp [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+    // The key should be same
+    ASSERT_EQ(it->key(), it_c->key());
+    // The modules should be the same kind (type).
+    ASSERT_EQ(it->value()->name(), it_c->value()->name());
+    // But not pointer-equal (distinct objects).
+    ASSERT_NE(it->value(), it_c->value());
+  }
+
+  // Verify that the clone is deep, i.e. parameters of modules are cloned too.
+  torch::NoGradGuard no_grad;
```
Comment (zou3519) [inline test/cpp/api/moduledict.cpp:221]: Why is this needed?

**Review by zou3519 — COMMENTED** [review]
Not many comments from me but I don't know much about the C++ API

File: torch/csrc/api/include/torch/nn/modules/container/moduledict.h (hunk 2 above)
Comment (glaringlee) [reply torch/csrc/api/include/torch/nn/modules/container/moduledict.h:97]: @zou3519 
Can not raise an error, since this is called in cloneable.h when cloning, even it is empty, it will be called. Leave it empty is fine.

File: torch/csrc/api/include/torch/nn/modules/container/moduledict.h [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+  /// Pretty prints the `ModuleDict` into the given `stream`.
+  void pretty_print(std::ostream& stream) const override {
+    stream << "torch::nn::ModuleDict";
+  }
+
+  /// Returns the value associated with the given `key`. Throws an exception if
+  /// no such key is stored in the `ModuleDict`. Check contains(key) before
+  /// for a non-throwing way of access.
+  std::shared_ptr<Module>& operator[](const std::string& key) {
```
Comment (glaringlee) [inline torch/csrc/api/include/torch/nn/modules/container/moduledict.h:107]: Why this is a reference instead of a shared_ptr obj?

File: torch/csrc/api/include/torch/nn/modules/container/moduledict.h [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    return modules_[key];
+  }
+
+  /// Returns the value associated with the given `key`. Throws an exception if
+  /// no such key is stored in the `ModuleDict`. Check contains(key) before
+  /// for a non-throwing way of access.
+  const std::shared_ptr<Module>& operator[](const std::string& key) const {
```
Comment (glaringlee) [inline torch/csrc/api/include/torch/nn/modules/container/moduledict.h:114]: Why this is a reference instead of a shared_ptr obj?

File: torch/csrc/api/include/torch/nn/modules/container/moduledict.h (hunk 2 above)
Comment (glaringlee) [reply torch/csrc/api/include/torch/nn/modules/container/moduledict.h:97]: @zou3519 I reviewed a bit, clone() is also overridden in this impl, so reset() is actually useless. I think it is fine to just leave it as blank, this is kinda convention in c++ API impls.

File: test/cpp/api/moduledict.cpp (hunk 4 above)
Comment (ejguan) [reply test/cpp/api/moduledict.cpp:221]: Yes, no need to run autograd for the following function as the test is to verify clone..

File: torch/csrc/api/include/torch/nn/modules/container/moduledict.h (hunk 5 above)
Comment (ejguan) [reply torch/csrc/api/include/torch/nn/modules/container/moduledict.h:107]: I am trying to keep consistency of our API like `parameterdict` and `orderedDict` for the operator [] and at function to return a value reference.

File: torch/csrc/api/include/torch/nn/modules/container/moduledict.h [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+    return *modules_[key]->as<T>();
+  }
+
+  /// Removes and returns the value associated with the given `key`.
+  /// Throws an exception if no such key is stored in the `ModuleDict`.
+  /// Check contains(key) before for a non-throwing way of access.
+  std::shared_ptr<Module> pop(const std::string& key) {
+    auto module = modules_[key];
+    modules_.erase(key);
+    return module;
```
[the remark this answers is not in the document]
Comment (ejguan) [reply torch/csrc/api/include/torch/nn/modules/container/moduledict.h:146]: It's same to `clear()`. I will also make a comment there.

File: torch/csrc/api/include/torch/nn/modules/container/moduledict.h (hunk 5 above)
Comment (ejguan) [reply torch/csrc/api/include/torch/nn/modules/container/moduledict.h:107]: Besides, by using reference, we can assign module to the location. But, it makes me concern about the registration. I think you are right, I need to make it non-reference obj.

**Comment by ejguan** [comment]
And, I am wondering if we need another API of `ptr(const std::string& key)`?  @glaringlee

File: torch/csrc/api/include/torch/nn/modules/container/moduledict.h (hunk 5 above)
Comment (glaringlee) [reply torch/csrc/api/include/torch/nn/modules/container/moduledict.h:107]: @ejguan Yea, there is no much difference here, a reference of shard_ptr is rarely being used (probably can save a counter++ time, I am not sure whether compiler will automatically invoke std::move, if so, then no overhead). Let's do non-reference obj.

**Comment by glaringlee** [comment]
> And, I am wondering if we need another API of `ptr(const std::string& key)`? @glaringlee

I did not see any usage for ptr at the moment, but you can add it if it is not heavy work.

File: test/cpp/api/moduledict.cpp [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+      std::dynamic_pointer_cast<ModuleDictImpl>(dict->clone(device));
+  for (const auto& p : clone->parameters()) {
+    ASSERT_EQ(p.device(), device);
+  }
+  for (const auto& b : clone->buffers()) {
+    ASSERT_EQ(b.device(), device);
+  }
+}
+
+TEST_F(ModuleDictTest, IsCloneable_CUDA) {
```
Comment (glaringlee) [inline test/cpp/api/moduledict.cpp:283]: It seems to me that IsCloneable and IsCloneable_CUDA are pretty much the same except the device.
Can we do something like this so we don't need to maintain 2 pieces of identical code. Sry, I should mention this earlier.
Here is an example:
https://github.com/pytorch/pytorch/blob/e70c50bf59f3ca0d885a716043e2f31aea98be30/test/cpp/api/transformer.cpp#L152

**Review by glaringlee — COMMENTED** [review]
Just one comment, I should mention earlier, sry.  
Otherwise, all looks good to me.

File: test/cpp/api/moduledict.cpp (hunk 8 above)
Comment (ejguan) [reply test/cpp/api/moduledict.cpp:283]: Yep, will do. Thanks for the advice!!

**Review by glaringlee — APPROVED** [review]
LGTM now. Please sync this latest version on phabricator before landing. thx
